File: frontend/src/utils/brief.ts
/**
 * The preview of the brief the events service writes for an event whose uploader did not write one.
 *
 * The convention itself lives in the events service, which is what actually names the event; this renders the
 * same parts in the same order so that the wizard can show what is about to be written rather than leaving
 * the user to press Create and find out. The one part it cannot know is the running number, which the service
 * mints when the event is stored.
 */

/** How the parts of a brief are joined, matching the convention of the events service exactly. */
const PART_SEPARATOR = ' · '
const PLATFORM_SEPARATOR = ' + '
const TYPE_SEPARATOR = ' / '

/** Past this the platforms are counted rather than named, so a brief stays a line somebody can read. */
const MAX_NAMED_PLATFORMS = 3

/** What an event whose type has not been picked yet is called while the preview is still incomplete. */
const UNTYPED_LABEL = 'Event'

/** How the running number the service mints is stood in for while the event does not have one yet. */
const PENDING_NUMBER = '#…'

/** The day spelled out rather than written in digits, so it cannot be mistaken for a reference number. */
const DATE_OPTIONS: Intl.DateTimeFormatOptions = { day: '2-digit', month: 'short', year: 'numeric' }

interface BriefParts {
  /** Names of the event types the event is filed under. */
  typeNames: string[]
  /** Keys of the platforms the event ran on. */
  platforms: string[]
  /** Key of the industry the event belongs to. */
  industry: string | null
  /** The day of the activity as the date input holds it, empty when the type does not ask for one. */
  eventDate: string
}

/**
 * Name the platforms an event ran on, counting them once there are more of them than a brief can carry.
 */
const platformLabel = (platforms: string[]): string => {
  const named = platforms.filter((platform) => platform.length > 0)
  if (named.length === 0) {
    return ''
  }

  if (named.length <= MAX_NAMED_PLATFORMS) {
    return named.join(PLATFORM_SEPARATOR)
  }

  const shown = named.slice(0, MAX_NAMED_PLATFORMS).join(PLATFORM_SEPARATOR)

  return `${shown} +${named.length - MAX_NAMED_PLATFORMS}`
}

/**
 * Read the day a brief is dated by, which is the day of the activity or, without one, today.
 */
const briefDate = (eventDate: string): string => {
  const moment = eventDate.length > 0 ? new Date(`${eventDate}T00:00:00`) : new Date()

  return Number.isNaN(moment.getTime())
    ? new Date().toLocaleDateString('en-GB', DATE_OPTIONS)
    : moment.toLocaleDateString('en-GB', DATE_OPTIONS)
}

/**
 * Render the brief the events service is about to write, as far as the wizard can know it.
 */
const buildEventBriefPreview = (parts: BriefParts): string => {
  const collected: string[] = [
    parts.typeNames.length > 0 ? parts.typeNames.join(TYPE_SEPARATOR) : UNTYPED_LABEL,
  ]

  const platforms = platformLabel(parts.platforms)
  if (platforms.length > 0) {
    collected.push(platforms)
  }

  if (parts.industry !== null && parts.industry.length > 0) {
    collected.push(parts.industry)
  }

  collected.push(briefDate(parts.eventDate))
  collected.push(PENDING_NUMBER)

  return collected.join(PART_SEPARATOR)
}

export type { BriefParts }
export { buildEventBriefPreview }

File: skyscanner/services/events_service/src/events_service/services/brief.py
"""
The convention behind an event brief that nobody wrote, built out of what the event already says about itself.

:date: 2026-08-27
:author: t_beatrice
"""
# ----- IMPORTS ----- #

from datetime import datetime

from skyscanner_models.enums import UploadSource

# ----- CONSTS ----- #

# The brief is read at a glance in a table of a hundred rows, so it is written as a few short facts rather
# than as a sentence: what happened, where it ran, which industry it belongs to, when, and which event it is.
PART_SEPARATOR: str = " · "
PLATFORM_SEPARATOR: str = " + "
TYPE_SEPARATOR: str = " / "

# The day is spelled out rather than written in digits, because a brief is read next to briefs somebody typed
# and "27 Aug 2026" cannot be mistaken for a reference number the way 27/08/2026 can.
DATE_FORMAT: str = "%d %b %Y"

# The running number of the event, which is what makes a generated brief unique even when two events of the
# same type ran on the same platform on the same day.
NUMBER_PREFIX: str = "#"

# Past this the platforms are counted rather than named, so that an event that ran on eleven rigs does not
# turn into a brief nobody can read.
MAX_NAMED_PLATFORMS: int = 3

# What an event without a single declared type is called, which only a caller that sent none can produce.
UNTYPED_LABEL: str = "Event"

# How the event reached the system, named only when it was not a person filling the wizard in - which is the
# ordinary case and therefore the one worth saying nothing about.
SOURCE_PREFIX: str = "via "

# ----- FUNCTIONS ----- #


def build_event_brief(
    type_names: list[str],
    platforms: list[str],
    industry: str,
    moment: datetime,
    event_number: int,
    upload_source: UploadSource = UploadSource.MANUAL,
) -> str:
    """
    Write the brief of an event whose uploader did not write one.

    The brief is what an event is listed, searched and recognised by, and asking for it was the one thing
    standing between a watchdog - or a user in a hurry - and an uploaded event. It is therefore derived
    instead, out of the four things every event already carries: what kind of activity it was, where it ran,
    which industry it belongs to and when it happened. The running number closes it, so two events that agree
    on all four are still told apart, and a brief that reads oddly is corrected on the event page like any
    other value - nothing about it is fixed once it has been written.

    :param type_names: Names of the event types the event was filed under.
    :param platforms: Keys of the platforms the event ran on.
    :param industry: Key of the industry the event belongs to.
    :param moment: Moment the brief is dated by, which is the date of the activity or of the upload.
    :param event_number: Running number the system minted for the event.
    :param upload_source: How the event reached the system.
    :return: The generated brief.
    """
    parts: list[str] = [TYPE_SEPARATOR.join(type_names) if type_names else UNTYPED_LABEL]

    platform_label = _platform_label(platforms=platforms)
    if platform_label:
        parts.append(platform_label)

    if industry:
        parts.append(industry)

    parts.append(moment.strftime(DATE_FORMAT))

    if upload_source is not UploadSource.MANUAL:
        parts.append(f"{SOURCE_PREFIX}{upload_source.value}")

    parts.append(f"{NUMBER_PREFIX}{event_number}")

    return PART_SEPARATOR.join(parts)


def _platform_label(platforms: list[str]) -> str:
    """
    Name the platforms an event ran on, counting them once there are more of them than a brief can carry.

    :param platforms: Keys of the platforms the event ran on.
    :return: The platforms as one readable piece of the brief, empty when the event named none.
    """
    named = [platform for platform in platforms if platform]
    if not named:
        return ""

    if len(named) <= MAX_NAMED_PLATFORMS:
        return PLATFORM_SEPARATOR.join(named)

    return f"{PLATFORM_SEPARATOR.join(named[:MAX_NAMED_PLATFORMS])} +{len(named) - MAX_NAMED_PLATFORMS}"
